### ISSUE TITLE
What is the different between train_loss and test_loss?

### ISSUE THREAD

**Issue by wulongjian** [pytorch/examples#600]
Hello, I am a student who is just beginning to learn pytorch, I have runnd the examples of MNIST code, I am curious why the train_loss and test_loss are calculated differently.Here is the calculation code.Why are they not using the same code?
`loss = F.nll_loss(output, target)  # train_loss`

` test_loss = F.nll_loss(output, target, reduction='sum')`
`test_loss = test_loss/len(test_loader.dataset)`

**Comment by luo3300612** [comment]
I think original code is
```python
for data, target in test_loader:
            ...
            test_loss += F.nll_loss(output, target, reduction='sum').item() # sum up batch loss
            ...
 test_loss /= len(test_loader.dataset)
```
`F.nll_loss` would compute the **average loss** of a batch by default when we train our model. Because we want to know how well each batch performs at this time to babysit our training process

when it's test time, we want to know the average loss on the whole test set. So we set `reduction='sum'` then it will compute the sum loss of a batch instead of average. Then we divide the size of test set `len(test_loader.dataset)` to get this average on the whole test set.

There are many ways to compute average loss on the whole test set, like:
```python
for data, target in test_loader:
            ...
            test_loss += F.nll_loss(output, target).item() * data.size(0) # average loss * batch_size
            ...
 test_loss /= len(test_loader.dataset)
```
